File: backend/app/core/periods.py
"""Period range helper — port of src/lib/date/periods.ts.

Dates are SERVER-LOCAL (IST), never UTC — see app/core/dates.py.
"""
from datetime import datetime, timedelta

from app.core.dates import to_iso_date


def get_period_range(period: str, now: datetime | None = None) -> dict:
    now = now or datetime.now()
    to = to_iso_date(now)

    if period == "week":
        return {"from": to_iso_date(now - timedelta(days=7)), "to": to, "label": "Last 7 days"}

    if period == "year":
        return {
            "from": to_iso_date(datetime(now.year, 1, 1)),
            "to": to,
            "label": f"Year {now.year}",
        }

    return {
        "from": to_iso_date(datetime(now.year, now.month, 1)),
        "to": to,
        "label": now.strftime("%B %Y"),
    }

File: frontend/src/lib/date/periods.ts
import { toISODate } from "./iso";

export type Period = "week" | "month" | "year" | "all";

// Earlier than any transaction this app can hold, so "all" needs no lower
// bound while still comparing as a plain ISO string like every other range.
const BEGINNING_OF_TIME = "1970-01-01";

export interface PeriodRange {
  from: string;
  to: string;
  label: string;
}

export function getPeriodRange(period: Period | string, now = new Date()): PeriodRange {
  const to = toISODate(now);

  if (period === "all") {
    return { from: BEGINNING_OF_TIME, to, label: "All time" };
  }

  if (period === "week") {
    const from = new Date(now);
    from.setDate(now.getDate() - 7);
    return { from: toISODate(from), to, label: "Last 7 days" };
  }

  if (period === "year") {
    return {
      from: toISODate(new Date(now.getFullYear(), 0, 1)),
      to,
      label: `Year ${now.getFullYear()}`,
    };
  }

  return {
    from: toISODate(new Date(now.getFullYear(), now.getMonth(), 1)),
    to,
    label: now.toLocaleString("en-IN", { month: "long", year: "numeric" }),
  };
}
